status/condition_set.go

```diff
@@ -0,0 +1,244 @@
+// on whether it is a Living or Batch process respectively.
+type ConditionTypes struct {
+	root       string
+	dependents []string
+}
+
+// NewReadyConditions returns a ConditionTypes to hold the conditions for the
+// resource. ConditionReady is used as the root condition.
+// The set of condition types provided are those of the terminal subconditions.
+func NewReadyConditions(d ...string) ConditionTypes {
+	return newConditionTypes(ConditionReady, d...)
+}
+
+// NewBatchConditions returns a ConditionTypes to hold the conditions for the
+// batch resource. ConditionSucceeded is used as the root condition.
+// The set of condition types provided are those of the terminal subconditions.
+func NewSucceededConditions(d ...string) ConditionTypes {
+	return newConditionTypes(ConditionSucceeded, d...)
+}
+
+func newConditionTypes(root string, dependents ...string) ConditionTypes {
+	return ConditionTypes{
+		root:       root,
+		dependents: lo.Reject(lo.Uniq(dependents), func(c string, _ int) bool { return c == root }),
+	}
+}
+
+// ConditionSet provides methods for evaluating Conditions.
+// +k8s:deepcopy-gen=false
+type ConditionSet struct {
+	ConditionTypes
+	object Object
+}
+
+// For creates a ConditionSet from an object using the original
+// ConditionTypes as a reference. Status must be a pointer to a struct.
+func (r ConditionTypes) For(object Object) ConditionSet {
+	cs := ConditionSet{object: object, ConditionTypes: r}
+	// Set known conditions Unknown if not set.
+	for _, t := range append(r.dependents, r.root) {
+		if cs.Get(t) == nil {
+			cs.SetUnknown(t)
+		}
+	}
+	return cs
+}
+
+// Root returns the root Condition, typically "Ready" or "Succeeded"
+func (c ConditionSet) Root() *Condition {
+	if c.object == nil {
+		return nil
+	}
+	return c.Get(c.root)
+}
+
+func (c ConditionSet) List() []Condition {
+	if c.object == nil {
+		return nil
+	}
+	return c.object.GetConditions()
+}
+
+// GetCondition finds and returns the Condition that matches the ConditionType
+// previously set on Conditions.
+func (c ConditionSet) Get(t string) *Condition {
+	if c.object == nil {
+		return nil
+	}
+	if condition, found := lo.Find(c.object.GetConditions(), func(c Condition) bool { return c.Type == t }); found {
+		return &condition
+	}
+	return nil
+}
+
+// Set sets or updates the Condition on Conditions for Condition.Type.
+// If there is an update, Conditions are stored back sorted.
+func (c ConditionSet) Set(condition Condition) {
+	if c.object == nil {
+		return
+	}
+	conditionType := condition.Type
+	var conditions []Condition
+	for _, c := range c.object.GetConditions() {
+		if c.Type != conditionType {
+			conditions = append(conditions, c)
+		} else {
+			// If we'd only update the LastTransitionTime, then return.
+			condition.LastTransitionTime = c.LastTransitionTime
+			if reflect.DeepEqual(condition, c) {
+				return
+			}
+		}
+	}
+	condition.LastTransitionTime = metav1.Now()
```

Comment:
This means that the lastTransitionTime will update even if the only thing that changes is the message? Is that the correct behavior?

Reply:
Seems fine to keep it this way for now